listener/src/serumListener.ts

```diff
@@ -0,0 +1,237 @@
+  const feeTier = event.feeTier
+
+  if (event.eventFlags.fill) {
+    const fillMessage: Fill = {
+      type: 'fill',
+      serumMarketAddress: market.address.toString(),
+      timestamp,
+      slot,
+      orderId,
+      clientId,
+      side,
+      price: _getFillPrice(market, event),
+      size: _getFillSize(market, event),
+      maker: event.eventFlags.maker,
+      feeCost: market.quoteSplSizeToNumber(event.nativeFeeOrRebate) * (event.eventFlags.maker ? -1 : 1),
+      account: openOrdersAccount,
+      accountSlot: openOrdersSlot,
+      feeTier: feeTier
+    }
+    return fillMessage
+  } else if (event.nativeQuantityPaid.eqn(0)) {
+    // we can use nativeQuantityPaid === 0 to detect if order is 'done'
+    // this is what the dex uses at event processing time to decide if it can release the slot in an OpenOrders account.
+    // done means that there won't be any more messages for the order (is no longer in the order book or never was - cancelled, ioc)
+
+    const doneMessage: Done = {
+      type: 'done',
+      serumMarketAddress: market.address.toString(),
+      timestamp,
+      slot,
+      orderId,
+      clientId,
+      side,
+      account: openOrdersAccount,
+      accountSlot: openOrdersSlot
+    }
+
+    return doneMessage
+  }
+
+  return
+}
+
+// TODO figure out how to make this resistant to concurrent calls for the same market 
+// (i.e. something like an advisory lock on a row, so we can determine if a market's 
+// event queue is being consumed by some other process)
+export const handleEventQueueChange = (connection: Connection, serumProgramId: PublicKey, market: Market) => async (accountInfo: AccountInfo<Buffer>, context: Context) => {
+  // retrieve the last event queue sequence number that was tracked from the database
+  const timestamp = new Date().toISOString()
+  const {error, response} = await getSerumMarketByAddress(market.address.toString())
+  if (error) return;
+  const { data } = await response.json()
+
+  if (data.serum_markets.length) {
+    const indexedMarket = data.serum_markets[0] as IndexedSerumMarket
+    let events: Event[] = [];
+
+    // Update the Serum market with the latest seq number
+    const header = EVENT_QUEUE_LAYOUT.HEADER.decode(accountInfo.data);
+    upsertSerumMarket({...indexedMarket, last_event_seq_num: header.seqNum})
+
+    if (indexedMarket.last_event_seq_num) {
+      events = decodeEventsSince(accountInfo.data, indexedMarket.last_event_seq_num)
+    } else {
+      events = decodeEventQueue(accountInfo.data)
+    }
+    // map the events to better structure
+    const formattedEvents: EventTypes[] = []
+    const fillsForTradeMatching: Record<string, Fill> = {}
+    events.forEach(event => {
+      const formattedEvent = _mapEventToDataMessage(event, market, timestamp, context.slot)
+      if (formattedEvent) {
+        formattedEvents.push(formattedEvent)
+        /**
+         * Below we capture a Trade event by matching opposite Fill events. This logic
```

Comment:
Is this safe? I'm not familiar enough with the order of events to have an opinion

Reply:
I'm not sure, but it worked the few times I tested. So I think we roll with it for now and maybe a better pattern will emerge. Plus as long as the _fills_ are in the database you can still figure out the trades